### PR TITLE
admin: improve infrastructure

### PR DESCRIPTION
**Pull request by trabetti**
Signed-off-by: trabetti <talis@il.ibm.com>

*title*: admin: add infrastructure, which will later be used for hystrix dashboard support 

*Description*: split from PR #2736
* Pass `AdminFilter` to admin handler, so that all its parameters can be used by the handler, without the need to pass them as arguments, will also enable adding more handler parameters without adding arguments.
* Handlers can add callbacks to be executed when connection is destroyed
* Support for streaming response

*Risk Level*: Medium 

*Testing*:
Will need to add coverage tests

### CODE REVIEW

File: source/server/http/admin.cc [hunk 1]
```diff
@@ -150,6 +150,16 @@ Http::FilterTrailersStatus AdminFilter::decodeTrailers(Http::HeaderMap&) {
   return Http::FilterTrailersStatus::StopIteration;
 }
 
+void AdminFilter::onDestroy() {
+  for (const auto& callback : on_destroy_callbacks_) {
```
Comment (jmarantz) [inline source/server/http/admin.cc]: for the record, generally we shouldn't use 'auto' as the type of a range variable when looping over a simple (non-map) container.  See

https://google.github.io/styleguide/cppguide.html#auto

However in this case it's OK because the usage makes it pretty obvious what you've got.

File: include/envoy/server/admin.h [hunk 2]
```diff
@@ -15,6 +15,7 @@
 namespace Envoy {
 namespace Server {
 
+class AdminFilter;
```
Comment (mattklein123) [inline include/envoy/server/admin.h]: nit: I'm not in favor of exposing out the AdminFilter object in this manner into the abstract includes. Can we add an `AdminFilter` or `AdminStream` interface here and then fill it in with relevant callbacks? Then the existing admin filter can implement this interface as needed?

Comment (mattklein123) [reply include/envoy/server/admin.h]: Either is fine with me.

Comment (trabetti) [reply include/envoy/server/admin.h]: @mattklein123 @jmarantz  - I pushed a fix. Not sure if what I did is what you meant here. Please review.

File: include/envoy/server/admin.h [hunk 3]
```diff
@@ -15,7 +16,14 @@
...
 namespace Server {
 
-class AdminFilter;
+class AdminStream {
+public:
+  virtual ~AdminStream() {}
+  virtual void setEndStreamOnComplete(const bool& end_stream) PURE;
+  virtual void addOnDestroyCallback(std::function<void()> cb) PURE;
+  virtual Http::StreamDecoderFilterCallbacks* getDecoderFilterCallbacks() PURE;
+  virtual const Http::HeaderMap* getRequestHeaders() PURE;
```
Comment (jmarantz) [inline include/envoy/server/admin.h]: `virtual const Http::HeaderMap& getRequestHeaders() const PURE;`

File: include/envoy/server/admin.h [hunk 4]
```diff
@@ -15,6 +16,14 @@
 namespace Envoy {
 namespace Server {
 
+class AdminStream {
+public:
+  virtual ~AdminStream() {}
+  virtual void setEndStreamOnComplete(const bool& end_stream) PURE;
```
Comment (mattklein123) [inline include/envoy/server/admin.h]: Can you doc comment all these methods?

File: include/envoy/server/admin.h [hunk 5]
```diff
@@ -15,7 +16,14 @@
 namespace Envoy {
 namespace Server {
 
-class AdminFilter;
+class AdminStream {
+public:
+  virtual ~AdminStream() {}
+  virtual void setEndStreamOnComplete(const bool& end_stream) PURE;
+  virtual void addOnDestroyCallback(std::function<void()> cb) PURE;
+  virtual Http::StreamDecoderFilterCallbacks* getDecoderFilterCallbacks() PURE;
```
Comment (jmarantz) [inline include/envoy/server/admin.h]: `virtual const Http::StreamDecoderFilterCallbacks& getDecoderFilterCallbacks() const PURE;`

return callbacks by const ref, and make the method const.

File: include/envoy/server/admin.h [hunk 6]
```diff
@@ -23,8 +32,8 @@ namespace Server {
  */
 #define MAKE_ADMIN_HANDLER(X)                                                                      \
   [this](absl::string_view path_and_query, Http::HeaderMap& response_headers,                      \
-         Buffer::Instance& data) -> Http::Code {                                                   \
-    return X(path_and_query, response_headers, data);                                              \
+         Buffer::Instance& data, AdminStream& admin_filter) -> Http::Code {                        \
+    return X(path_and_query, response_headers, data, admin_filter);                                \
```
Comment (mattklein123) [inline include/envoy/server/admin.h]: nit: s/admin_filter/admin_stream here and elsewhere

File: source/server/http/admin.h [hunk 7]
```diff
@@ -228,12 +235,14 @@ class AdminImpl : public Admin,
 /**
  * A terminal HTTP filter that implements server admin functionality.
  */
-class AdminFilter : public Http::StreamDecoderFilter, Logger::Loggable<Logger::Id::admin> {
+class AdminFilter : public Http::StreamDecoderFilter,
+                    AdminStream,
```
Comment (mattklein123) [inline source/server/http/admin.h]: public AdminStream

File: include/envoy/server/admin.h [hunk 8]
```diff
@@ -15,6 +16,34 @@
 namespace Envoy {
 namespace Server {
 
+class AdminStream {
+public:
+  virtual ~AdminStream() {}
+
+  /**
+   * @param end_stream set to false for streaming response. Default is true, which will close
+   * connection on socket
```
Comment (jmarantz) [inline include/envoy/server/admin.h]: End sentences with period, here & below.

File: include/envoy/server/admin.h [hunk 9]
```diff
@@ -15,6 +16,34 @@
...
+   * @param cb callback to be added to the list of callbacks invoked by onDestroy() when connection
+   * is dropped
+   */
+  virtual void addOnDestroyCallback(std::function<void()> cb) PURE;
+
+  /**
+   * @return Http::StreamDecoderFilterCallbacks* to be used by the handler to get socket for data
+   * streaming
+   */
+  virtual const Http::StreamDecoderFilterCallbacks* getDecoderFilterCallbacks() const PURE;
```
Comment (jmarantz) [inline include/envoy/server/admin.h]: Can you either make a guarantee that the return-value is non-null by returning a ref, or documenting that nullptr might be an actual return-value?

I think that in the actual usage model, there's no way for null to be returned in the context of an admin callback to which the stream object is passed.  However, I can see that AdminFilter does have these null on construction.

So what's the correct way to use this, and getRequestHeaders()?

One idea is you can make this type-safe by having an StreamImpl object which contains a pointer to the AdminFilter as a delegate, and implements these methods with guarantees implied by returning const ref for the callbacks and requestHeaders, know that at the time is is called, those are guaranteed non-null.

In that case you wouldn't multiply inherit from AdminFilter to AdminStream.

This is a slightly more complex implementation, but a much simpler-to-describe behavior.  With you have, I think you can't make the guarantee without complicated language, and then the caller needs to check for nullptr, creating hard-to-test paths through the code.

WDYT?

Comment (trabetti) [reply include/envoy/server/admin.h]: I do not feel I am familiar enough with the behavior to decide whether we can guarantee a non-null.  Maybe someone else can comment on this. Can you be more explicit about the suggested solution?

Comment (trabetti) [reply include/envoy/server/admin.h]: Another question - there is coverage missing here, but most of it will be covered by the Hystrix test. Can we skip the coverage testing now?

Comment (jmarantz) [reply include/envoy/server/admin.h]: RE guarantees: look at the call-site for runCallback in AdminFilter::onComplete, and you'll see that there has to be non-null request-headers there (there's a RELEASE_ASSERT).  At that point you could create the AdminStream object on the stack with the non-null request-headers ref and pass it to runCallback, rather than 'this' (having cut the multiple-inheritance relationship).

It looks to me also that callbacks_ has to be non-null at that point as well, or the call to encodeHeaders will segv.

Comment (jmarantz) [reply include/envoy/server/admin.h]: re code-coverage, ideally each level of abstraction should have its own covering unit tests, and not rely on some other complex system to test it.  Is there some reason why this would be hard to test in this case?

Comment (jmarantz) [reply include/envoy/server/admin.h]: No, I suggest: 
  1. have AdminStream be a pure base class, not part of a multi-inheritance mesh.
  2. have the callbacks & req-headers be const refs member vars of AdminStreamImpl
  3. construct AdminStreamImpl in AdminFilters::onComplete where it's clear you have those
      available non-null.

Comment (trabetti) [reply include/envoy/server/admin.h]: Re code-coverage: it will be hard to test the onDestroy() code with added callbacks. I could use addHandler() to add a new handler that adds a callback. The two relevant tests are admin_test and integration_admin_test. In admin_test I have an admin object but no requests, so onDestroy() is not called.  In integration_admin_test onDestroy() is called but I don't have an admin object.
I may be missing a simple solution.. it looks to me that it is complicated to build such test, but hystrix_test should cover it, that's why I asked.

Comment (jmarantz) [reply include/envoy/server/admin.h]: I understand there are not already existing tests, but I'm still not clear why it's difficult to make one.

I definitely think it's better to cover each level of abstraction on its own and not wait for its application with a higher level integration to be able to test it.

File: include/envoy/server/instance.h [hunk 10]
```diff
@@ -179,6 +179,11 @@ class Instance {
    * @return information about the local environment the server is running in.
    */
   virtual const LocalInfo::LocalInfo& localInfo() PURE;
+
+  /**
+   * @return the flush interval of stats sinks.
+   */
+  virtual std::chrono::milliseconds statsFlushInterval() PURE;
```
Comment (jmarantz) [inline include/envoy/server/instance.h]: const method, here &  the override

File: test/integration/integration_admin_test.cc [hunk 11]
```diff
@@ -266,6 +266,56 @@ TEST_P(IntegrationAdminTest, Admin) {
...
+//      test_server_->server().admin().addHandler("/foo/bar", "hello", callback, true, false));
+//
+//  BufferingStreamDecoderPtr response = IntegrationUtil::makeSingleRequest(
+//      lookupPort("admin"), "GET", "/foo/bar", "", downstreamProtocol(), version_);
+//
+//  EXPECT_TRUE(response->complete());
+//  EXPECT_STREQ("200", response->headers().Status()->value().c_str());
+//  EXPECT_EQ(spdlog::level::trace, Logger::Registry::getLog(Logger::Id::assert).level());
+//  // TODO (@trabetti) : what can be checked on response headers?
+//}
```
Comment (jmarantz) [inline test/integration/integration_admin_test.cc]: What's the story here?

File: include/envoy/server/instance.h (hunk 10 above)
Comment (trabetti) [reply include/envoy/server/instance.h]: The implementation calls statsFlushInterval() of config, which is not a const ref, would you change there too?
https://github.com/envoyproxy/envoy/blob/ed468c9522d6dda82e15c7ba1cc3cb2204ada81b/include/envoy/server/configuration.h#L76

File: test/integration/integration_admin_test.cc (hunk 11 above)
Comment (trabetti) [reply test/integration/integration_admin_test.cc]: tried to see if there's a way to test end_stream = false. but then the response never returns.

File: include/envoy/server/configuration.h [hunk 12]
```diff
@@ -73,7 +73,7 @@ class Main {
    * @return std::chrono::milliseconds the time interval between flushing to configured stat sinks.
    *         The server latches counters.
    */
-  virtual std::chrono::milliseconds statsFlushInterval() PURE;
+  virtual const std::chrono::milliseconds& statsFlushInterval() PURE;
```
Comment (jmarantz) [inline include/envoy/server/configuration.h]: why return a const ref here?  Most of the other time-returning things I see in Envoy are returning by value.

File: source/server/config_validation/server.h [hunk 13]
```diff
@@ -89,6 +89,10 @@ class ValidationInstance : Logger::Loggable<Logger::Id::main>,
   ThreadLocal::Instance& threadLocal() override { return thread_local_; }
   const LocalInfo::LocalInfo& localInfo() override { return *local_info_; }
 
+  const std::chrono::milliseconds& statsFlushInterval() override {
```
Comment (jmarantz) [inline source/server/config_validation/server.h]: const  method per discussion in outdated conversation.  You will have to fix the underlying interface and impl.

not const ref return though.

File: include/envoy/server/instance.h [hunk 14]
```diff
@@ -179,6 +179,11 @@ class Instance {
    * @return information about the local environment the server is running in.
    */
   virtual const LocalInfo::LocalInfo& localInfo() PURE;
+
+  /**
+   * @return the flush interval of stats sinks.
+   */
+  virtual const std::chrono::milliseconds& statsFlushInterval() PURE;
```
Comment (jmarantz) [inline include/envoy/server/instance.h]: const method; not const ref return.

File: source/server/configuration_impl.h [hunk 15]
```diff
@@ -127,7 +127,7 @@ class MainImpl : Logger::Loggable<Logger::Id::config>, public Main {
   Tracing::HttpTracer& httpTracer() override { return *http_tracer_; }
   RateLimit::ClientFactory& rateLimitClientFactory() override { return *ratelimit_client_factory_; }
   std::list<Stats::SinkPtr>& statsSinks() override { return stats_sinks_; }
-  std::chrono::milliseconds statsFlushInterval() override { return stats_flush_interval_; }
+  const std::chrono::milliseconds& statsFlushInterval() override { return stats_flush_interval_; }
```
Comment (jmarantz) [inline source/server/configuration_impl.h]: const ref, not const ref return.

File: source/server/http/admin.h [hunk 16]
```diff
@@ -31,6 +31,30 @@
 namespace Envoy {
 namespace Server {
 
+/**
+ * Implementation of Server::AdminStream.
+ */
+class AdminStreamImpl : public AdminStream {
+public:
+  AdminStreamImpl(Http::StreamDecoderFilterCallbacks& callbacks, Http::HeaderMap& request_headers,
```
Comment (jmarantz) [inline source/server/http/admin.h]: It's worth commenting here that the ctor args are held by reference and must be held in a data structure that outlasts the stream.  Another thought I had was to have the AdminFilter *own* the AdminStreamImpl as a std::unique_ptr, which and this would be the only location for these objects.  That might be easier to reason about in terms of correctness.  It would  populate the AdminStreamImpl in the same place it does now.

File: test/server/http/admin_test.cc [hunk 17]
```diff
@@ -206,30 +216,68 @@ TEST_P(AdminInstanceTest, CustomHandler) {
...
+                     AdminStream& admin_stream) -> Http::Code {
+    admin_stream.setEndStreamOnComplete(false);
+
+    // check that value is set to false
+    if (admin_stream.endStreamOnComplete())
+      return Http::Code::NotAcceptable;
+    return Http::Code::OK;
   };
+
+  // Add non removable handler.
```
Comment (jmarantz) [inline test/server/http/admin_test.cc]: it would be nicer to make separate testcases for these.  You could generate actually make the callback be a static method of AdminInstanceTest to share it between the methods.

File: source/server/http/admin.h (hunk 16 above)
Comment (trabetti) [reply source/server/http/admin.h]: ok, will look into it, I just need to check with the Hystrix implementation that it doesn't break anything

File: test/server/http/admin_test.cc (hunk 17 above)
Comment (trabetti) [reply test/server/http/admin_test.cc]: I thought it is better to cover all these setters/getters in one test.

File: source/server/config_validation/server.h (hunk 13 above)
Comment (trabetti) [reply source/server/config_validation/server.h]: Sure, I should have paid attention. will correct all these.

File: test/server/http/admin_test.cc (hunk 17 above)
Comment (jmarantz) [reply test/server/http/admin_test.cc]: Common practice is to make each individual test shorter.

File: test/server/http/admin_test.cc [hunk 18]
```diff
@@ -206,30 +216,68 @@ TEST_P(AdminInstanceTest, CustomHandler) {
...
+
+  // Cover getRequestHeaders() method.
+  const Http::TestHeaderMapImpl temp_request_headers;
+
+  const Http::TestHeaderMapImpl got_request_headers = admin_stream.getRequestHeaders();
+  EXPECT_EQ(got_request_headers.byteSize(), request_headers_.byteSize());
+}
+
+TEST_P(AdminInstanceTest, RejectHandlerWithXss) {
+  auto callback = [](absl::string_view, Http::HeaderMap&, Buffer::Instance&,
```
Comment (jmarantz) [inline test/server/http/admin_test.cc]: You could factor this helper callback generator into a simple static test method, since it is used 3x (previously 2x)

File: source/server/http/admin.h [hunk 19]
```diff
@@ -252,6 +283,8 @@ class AdminFilter : public Http::StreamDecoderFilter, Logger::Loggable<Logger::I
   AdminImpl& parent_;
   Http::StreamDecoderFilterCallbacks* callbacks_{};
   Http::HeaderMap* request_headers_{};
+  std::unique_ptr<std::list<std::function<void()>>> on_destroy_callbacks_ =
```
Comment (mattklein123) [inline source/server/http/admin.h]: nit: I don't think it's really worth optimizing this to create the list when needed, as this is admin code. I would just remove the unique_ptr here and always have the list.

File: include/envoy/server/admin.h [hunk 20]
```diff
@@ -15,6 +16,39 @@
 namespace Envoy {
 namespace Server {
 
+class AdminStream {
+public:
+  virtual ~AdminStream() {}
+
+  /**
+   * @param end_stream set to false for streaming response. Default is true, which will close
```
Comment (mattklein123) [inline include/envoy/server/admin.h]: "which will close connection on socket" is misleading and not necessarily true if the admin connection were kept alive, which is possible with the existing code. I would rephrase to make it clear that this is really about ending the response when the initial handler completes or something like that.

File: include/envoy/server/admin.h [hunk 21]
```diff
@@ -15,6 +16,39 @@
...
+   * connection on socket.
+   */
+  virtual void setEndStreamOnComplete(bool end_stream) PURE;
+
+  /**
+   * @param cb callback to be added to the list of callbacks invoked by onDestroy() when connection
+   * is dropped.
```
Comment (mattklein123) [inline include/envoy/server/admin.h]: Same comment. Connection is not necessarily dropped. When the stream is closed is more accurate.

File: include/envoy/server/admin.h [hunk 22]
```diff
@@ -15,6 +16,39 @@
...
+   */
+  virtual void addOnDestroyCallback(std::function<void()> cb) PURE;
+
+  /**
+   * @return Http::StreamDecoderFilterCallbacks* to be used by the handler to get socket for data
```
Comment (mattklein123) [inline include/envoy/server/admin.h]: We are not really getting socket data here. We are getting HTTP request data, etc. I would rephrase.

File: source/server/http/admin.h (hunk 19 above)
Comment (trabetti) [reply source/server/http/admin.h]: Sure? Not sure it will have any use beside in the Hystrix handler.

Comment (mattklein123) [reply source/server/http/admin.h]: I don't feel very strongly about it, but this makes the code harder to read and I don't think the slightly smaller CPU/memory use matters here.

File: include/envoy/server/admin.h [hunk 23]
```diff
@@ -15,6 +16,39 @@
...
+   * @param end_stream set to false for streaming response. Default is true, which will
+   * end the response when the initial handler completes.
+   */
+  virtual void setEndStreamOnComplete(bool end_stream) PURE;
+
+  /**
+   * @param cb callback to be added to the list of callbacks invoked by onDestroy() when connection
+   * is closed.
+   */
+  virtual void addOnDestroyCallback(std::function<void()> cb) PURE;
```
Comment (jmarantz) [inline include/envoy/server/admin.h]: nit/taste: I feel like the 'On' here and the 'Callback' both add the same info, and it would be slightly nicer to call this 'addDestroyCallback', but I'm fine as it is too.

File: source/server/http/admin.h [hunk 24]
```diff
@@ -31,6 +31,30 @@
...
+  void setEndStreamOnComplete(bool end_stream) override { end_stream_on_complete_ = end_stream; }
+  void addOnDestroyCallback(std::function<void()> cb) override;
+  const Http::StreamDecoderFilterCallbacks& getDecoderFilterCallbacks() const override {
+    return callbacks_;
+  }
+  const Http::HeaderMap& getRequestHeaders() const override { return request_headers_; }
+  bool endStreamOnComplete() const override { return end_stream_on_complete_; }
+
+private:
+  const Http::StreamDecoderFilterCallbacks& callbacks_;
```
Comment (jmarantz) [inline source/server/http/admin.h]: comment here too about who owns this, if you prefer not to try to have AdminStream be the sole owner (with AdminFilter owning the AdminStream as a std::unique_ptr) as I suggested above.

File: include/envoy/server/admin.h (hunk 23 above)
Comment (trabetti) [reply include/envoy/server/admin.h]: thanks for the comment. I prefer to keep the name since it references the onDestroy() method, where the callbacks are being used.

File: source/server/http/admin.h (hunk 24 above)
Comment (trabetti) [reply source/server/http/admin.h]: Verified this against the intended hystrix addition. working. pushing a fix.

File: source/server/http/admin.cc [hunk 25]
```diff
@@ -664,8 +664,9 @@ void AdminFilter::onComplete() {
   Http::HeaderMapPtr header_map{new Http::HeaderMapImpl};
   RELEASE_ASSERT(request_headers_);
 
-  AdminStreamImpl admin_stream(*callbacks_, *request_headers_, on_destroy_callbacks_);
-  Http::Code code = parent_.runCallback(path, *header_map, response, admin_stream);
+  std::unique_ptr<AdminStreamImpl> admin_stream =
+      std::make_unique<AdminStreamImpl>(*callbacks_, *request_headers_, on_destroy_callbacks_);
+  Http::Code code = parent_.runCallback(path, *header_map, response, *admin_stream);
```
Comment (jmarantz) [inline source/server/http/admin.cc]: I think my suggestion was not clear.  I was suggesting you make this unique_ptr<AdminStreamImpl> be a member variable of AdminFilter, and should own the headers, callbacks, and destroy-callbacks exclusively.

The existing of this as an independent class referencing member variables of AdminStream seemed perilous and I think that change makes the ownership semantics much clearer and easy to reason about, without affecting the actual flow of the logic.

WDYT?

Comment (trabetti) [reply source/server/http/admin.cc]: I don't mind this change.  But need to check if callbacks can be on a unique pointer,  since it points to outside admin.

**Comment by jmarantz** [comment]
In general I feel that holding onto refs to member vars of other classes is hard to reason about.  But it we were already doing that and your PR doesn't make it worse, then for callbacks it makes sense to leave then as a ref, but still doing as I suggested for the objects created in runCallback.

File: source/server/http/admin.h (hunk 16 above)
Comment (mrice32) [reply source/server/http/admin.h]: I'm not super opinionated about who needs to own the admin stream since either way it will be possible to be left with a dangling reference without explicitly destroying something IIUC (in once case it will be the stream, in the other it will be the callbacks).  However, as @jmarantz suggests, in either case, please document how one would know that the reference is invalidated (using `onDestryCallabacks()`).

**Comment by trabetti** [comment]
@jmarantz : I still think of the AdminStream as a compact way to pass parameters to and from the handlers. Making it a class member and moving members used by AdminFilter to be owned by the AdminStream seems to me more confusing. 
@mrice32 : a comment would be that the callbacks are invalidated after onDestory() is called, so anyone using them, should add a callback function to onDestroyCallbacks(), to invalidate whatever their handler is doing. This is what we are doing in the hystrix code. Is this what you mean, or is there anything we can do in onDestroy() to handle this in the general case?

**Comment by jmarantz** [comment]
Any time you have a data structure that contains references or pointers to member variables owned by other data structures, it casts suspicion and doubt and forces the reader to look very carefully.

That is why I believe clean ownership semantics are better, and less confusing.

I also suggested in this PR to call this 'RequestContext' rather than 'AdminStream', and it might grow into a useful abstraction beyond this over time.  In many other server architectures I've used in the past, a request-context is where this sort of data would live. But I don't have the breadth of knowledge of Envoy's structures yet to know how that would fit in more broadly.

However as I said initially when I made this suggestion, I'm also OK with commenting the lifetime of what each member variable is referring to is, particularly because as @mrice32 points out, at least one of the member variables would need lifetime beyond the AdminStream.

File: source/server/http/admin.cc [hunk 26]
```diff
@@ -653,7 +663,11 @@ void AdminFilter::onComplete() {
   Buffer::OwnedImpl response;
   Http::HeaderMapPtr header_map{new Http::HeaderMapImpl};
   RELEASE_ASSERT(request_headers_);
-  Http::Code code = parent_.runCallback(path, *request_headers_, *header_map, response);
+
+  std::unique_ptr<AdminStreamImpl> admin_stream =
```
Comment (jmarantz) [inline source/server/http/admin.cc]: My suggestion to make this a unique_ptr rather than instantiate it on-stack only makes sense if you are going to hold the unique_ptr in the AdminStream, moving the ownership of header_map and response along with it.

If you really don't want to do that, you might as well just instantiate the AdminStream on the stack here.  You are not releasing it or std::move()ing it either, so unique_ptr locally here doesn't have much impact.

File: include/envoy/server/admin.h [hunk 27]
```diff
@@ -15,6 +16,39 @@
...
+  virtual ~AdminStream() {}
+
+  /**
+   * @param end_stream set to false for streaming response. Default is true, which will
+   * end the response when the initial handler completes.
+   */
+  virtual void setEndStreamOnComplete(bool end_stream) PURE;
+
+  /**
+   * @param cb callback to be added to the list of callbacks invoked by onDestroy() when connection
```
Comment (mattklein123) [inline include/envoy/server/admin.h]: I can't remember if I commented on this before, but this is not getting called with the connection is closed, it's getting called when the stream is closed. Can you update the comment?

File: source/server/http/admin.cc [hunk 28]
```diff
@@ -653,7 +663,10 @@ void AdminFilter::onComplete() {
   Buffer::OwnedImpl response;
   Http::HeaderMapPtr header_map{new Http::HeaderMapImpl};
   RELEASE_ASSERT(request_headers_);
-  Http::Code code = parent_.runCallback(path, *request_headers_, *header_map, response);
+
+  AdminStreamImpl admin_stream(*callbacks_, *request_headers_, on_destroy_callbacks_);
```
Comment (mattklein123) [inline source/server/http/admin.cc]: I'm sorry, I haven't been tracking this lengthy conversation, but why doesn't AdminFilter implement AdminStream and then you just pass `*this`? That is the correct ownership model anyway IMO and avoids all of the other concerns? cc @jmarantz

File: source/server/http/admin.h [hunk 29]
```diff
@@ -49,6 +49,9 @@ class AdminStreamImpl : public AdminStream {
   bool endStreamOnComplete() const override { return end_stream_on_complete_; }
 
 private:
+  // the callbacks member is a reference to a member of a ConnectionManagerImpl object. It is
+  // destroyed when connection is terminated. handlers using it should invalidate the reference by
+  // adding a method to be run when onDestroy() is invoked, using addOnDestroyCallback().
   const Http::StreamDecoderFilterCallbacks& callbacks_;
   const Http::HeaderMap& request_headers_;
```
Comment (jmarantz) [inline source/server/http/admin.h]: similar comments are also needed for request_headers_ and on_destroy_callbacks_.  Basically any time you have a reference or pointer as a member variable you need to comment on ownership & lifetime.

File: source/server/http/admin.cc (hunk 28 above)
Comment (trabetti) [reply source/server/http/admin.cc]: We started with that model and then changed it in response to feedback. If it makes more sense we can go back to that. @mattklein123 @jmarantz

File: source/server/http/admin.h [hunk 30]
```diff
@@ -31,6 +31,33 @@
...
+  void setEndStreamOnComplete(bool end_stream) override { end_stream_on_complete_ = end_stream; }
+  void addOnDestroyCallback(std::function<void()> cb) override;
+  const Http::StreamDecoderFilterCallbacks& getDecoderFilterCallbacks() const override {
+    return callbacks_;
+  }
+  const Http::HeaderMap& getRequestHeaders() const override { return request_headers_; }
+  bool endStreamOnComplete() const override { return end_stream_on_complete_; }
+
+private:
+  // the callbacks member is a reference to a member of a ConnectionManagerImpl object. It is
```
Comment (mrice32) [inline source/server/http/admin.h]: nit: `The callbacks`.  Please capitalize the first letter of each sentence here and later in this comment.

File: source/server/http/admin.h [hunk 31]
```diff
@@ -31,6 +31,33 @@
...
+  // destroyed when connection is terminated. handlers using it should invalidate the reference by
```
Comment (mrice32) [inline source/server/http/admin.h]: nit: maybe reword the second sentence to to:
```
Handlers relying on the reference should use addOnDestroyCallback()
to add a callback that will notify them when the reference is no
longer valid.
```

File: source/server/http/admin.cc (hunk 28 above)
Comment (mattklein123) [reply source/server/http/admin.cc]: Alright, I don't mean to rehash a huge conversation. Personally, I think AdminFilter implementing AdminStream is the clearest model, as the filter/stream is alive for the lifetime of the admin request, so that is what I would do, but if we want to start with this we can see how it goes. Though I don't see how this can be the final solution as being stack allocated, as for handlers that continue processing they will do so outside the context of onComplete, so something will need to change?

Comment (trabetti) [reply source/server/http/admin.cc]: I don't know if it answers your question, but - In the Hystrix example, only the StreamDecoderFilterCallbacks is needed after onComplete is done with. It references to outside the AdminFilter. In Hystrix we keep pointer to it in the sink, so the pointer stay valid after the lifetime of onComplete(). The idea we agreed here is to handle this by adding methods to on_destroy_callbacks_ that will stop using the StreamDecoderFilterCallbacks pointer when onDestroy() is called. Adding this function should be done in the handler.

Comment (mattklein123) [reply source/server/http/admin.cc]: I've already commented on this several times, but there is nothing happening here at the granularity of the connection. Only streams, and the filter object has the lifetime of the stream. Trust me, you want the filter to implement the interface.

Comment (mrice32) [reply source/server/http/admin.cc]: +1.  Having the filter implement the interface should work perfectly for the Hystrix case.  It will make the lifetime easier to understand, and it will incur the smallest code footprint.

Comment (mattklein123) [reply source/server/http/admin.cc]: I talked about this offline with @jmarantz. I think we are all on the same page to move to the inheritance model. Sorry for the runaround @trabetti but I think important learning/exploration for all involved.

**Comment by trabetti** [comment]
@mattklein123 Just pushed a version with AdminFilter implementing AdminStream, is this what you meant?

**Comment by mattklein123** [comment]
@trabetti yes. Can you merge master and look at the remaining comments? Then I think we can get this merged.

**Comment by trabetti** [comment]
that would be great

**Comment by jmarantz** [comment]
Confirmed; I'm fine with going back to the multiple-inheritance model; but returning references to the objects that must not be nullptr, using the pattern
```
ASSERT(request_headers_.get() != nullptr);
return *request_headers_;
```

File: include/envoy/server/admin.h [hunk 32]
```diff
@@ -15,6 +16,34 @@
...
+  virtual void setEndStreamOnComplete(bool end_stream) PURE;
+
+  /**
+   * @param cb callback to be added to the list of callbacks invoked by onDestroy() when stream
+   * is closed.
+   */
+  virtual void addOnDestroyCallback(std::function<void()> cb) PURE;
+
+  /**
+   * @return Http::StreamDecoderFilterCallbacks* to be used by the handler to get HTTP request data
```
Comment (mattklein123) [inline include/envoy/server/admin.h]: nit: `Http::StreamDecoderFilterCallbacks&`

File: include/envoy/server/admin.h [hunk 33]
```diff
@@ -15,6 +16,34 @@
...
+   * for streaming.
+   */
+  virtual const Http::StreamDecoderFilterCallbacks& getDecoderFilterCallbacks() const PURE;
+
+  /**
+   * @return Http::HeaderMap* to be used by handler to parse header information sent with the
```
Comment (mattklein123) [inline include/envoy/server/admin.h]: nit: `Http::HeaderMap&`

**Comment by trabetti** [comment]
any idea what is this tsan failure?

**Comment by mattklein123** [comment]
@trabetti unrelated. Fixed in master.
